lib/core/src/Cardano/Wallet/Api/Types.hs

```diff
@@ -658,6 +681,39 @@ instance FromJSON WalletPostData where
 instance ToJSON  WalletPostData where
     toJSON = genericToJSON defaultRecordTypeOptions
 
+instance FromJSON AccountPublicKey where
+    parseJSON bytes = do
+        bs <- T.encodeUtf8 <$> parseJSON bytes
+        case xpubFromText bs of
+            Left _ ->
+                fail "AccountPublicKey: unable to deserialize ShelleyKey from json"
+            Right pubkey ->
+                pure $ AccountPublicKey $ ApiT $ ShelleyKey pubkey
+instance ToJSON AccountPublicKey where
+    toJSON =
+        toJSON . T.decodeUtf8 . serializeXPub . getApiT . key
+
+instance FromJSON WalletOrAccountPostData where
+    parseJSON obj = do
+        val <-
+            (withObject "postData" $
+             \o -> o .:? "passphrase" :: Aeson.Parser (Maybe Text)) obj
+        case val of
+            Nothing -> do
+                xs <- parseJSON obj :: Aeson.Parser AccountPostData
+                pure $ WalletOrAccountPostData $ Right xs
+            Just _ -> do
+                xs <- parseJSON obj :: Aeson.Parser WalletPostData
+                pure $ WalletOrAccountPostData $ Left xs
```

Comment:
Tips: parsers are alternatives. So we can simply write:

```hs
parseJSON obj = WalletOrAccountPostData <$>
    (Left <$> parseJSON obj <|> Right <$> parseJSON obj)
```

;)

Reply:
Could it not potentially result in worse error messages?

If there is a slight mistake in the `Left` branch, will it not continue with the `Right` branch and might produce an error message specific to the right branch?

Reply:
That's pretty much the case now too. If you forget the `passphrase` (or mistype it), we'll just assume an `AccountPostData` :man_shrugging:

We can probably do something more elaborate like we did for `FromMnemonic`. My two cents would actually be that, we should probably collect the error message from both branches and provide both to the user in a readable way.

Reply:
done

Reply:
hmmm, I have a number of things like that : 
```

1) API Specifications, WALLETS_CREATE_05 - Mnemonics, [] as mnemonic_sentence -> fail
--
  | "Error in $: key \"id\" not present: Response {responseStatus = Status {statusCode = 400, statusMessage = \"Bad Request\"}, responseVersion = HTTP/1.1, responseHeaders = [(\"Transfer-Encoding\",\"chunked\"),(\"Date\",\"Mon, 24 Feb 2020 19:49:06 GMT\"),(\"Server\",\"Warp/3.2.27\"),(\"Content-Type\",\"application/json;charset=utf-8\")], responseBody = \"{\\\"code\\\":\\\"bad_request\\\",\\\"message\\\":\\\"Error in $: key 'account_public_key' not present\\\"}\", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}" does not contain "Invalid number of words: 15, 18, 21 or 24 words are expected."
```
will need to work on error management in this instance